### PR TITLE
[Refactor] refactor dict filter logic in group reader

### PR DESCRIPTION
**Pull request by dirtysalt**
## What type of PR is this：
- [ ] BugFix
- [ ] Feature
- [ ] Enhancement
- [x] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #

## Problem Summary(Required) ：

This PR is to refactor dict filter logic: wrap dict filter related code into `DictFilterContext`.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto backported to target branch
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3
  - [ ] 2.2

### CODE REVIEW

File: be/src/formats/parquet/group_reader.cpp [hunk 1]
```diff
@@ -78,11 +78,11 @@ Status GroupReader::get_next(vectorized::ChunkPtr* chunk, size_t* row_count) {
...
     DCHECK_EQ(active_chunk->num_rows(), count);
 
-    // dict filter
-    if (has_dict_filter) {
+    // dict filter chunk
+    if (_dict_filter_ctx.has_predicates()) {
         SCOPED_RAW_TIMER(&_param.stats->expr_filter_ns);
         SCOPED_RAW_TIMER(&_param.stats->group_dict_filter_ns);
-        _dict_filter(&active_chunk, &chunk_filter);
+        _dict_filter_ctx.filter_chunk(&active_chunk, &chunk_filter);
```
Comment (DorianZheng) [inline be/src/formats/parquet/group_reader.cpp:85]: It's more simple if we let `filter_chunk` return bool value to indicate if it has predicates, then we don't need `has_predicates`

File: be/src/formats/parquet/group_reader.cpp [hunk 2]
```diff
@@ -179,25 +179,20 @@ Status GroupReader::_create_column_reader(const GroupReaderParam::Column& column
     return Status::OK();
 }
 
-void GroupReader::_process_columns_and_conjunct_ctxs() {
+Status GroupReader::_process_columns_and_conjunct_ctxs() {
```
Comment (DorianZheng) [inline be/src/formats/parquet/group_reader.cpp:182]: It's always return `Status::OK()`. It seems OK to remain returning `void` type

File: be/src/formats/parquet/group_reader.h [hunk 3]
```diff
@@ -85,18 +85,46 @@ class GroupReader {
...
+    private:
+        // if this column use as dict filter?
+        std::vector<bool> _is_dict_filter_column;
+        // columns(index) use as dict filter column
+        std::vector<int> _dict_column_indices;
+        // conjunct ctxs for each dict filter column
+        std::unordered_map<SlotId, std::vector<ExprContext*>> _conjunct_ctxs_by_slot;
+        // preds transformed from `_conjunct_ctxs_by_slot` for each dict filter column
+        std::unordered_map<SlotId, vectorized::ColumnPredicate*> _predicates;
+        GroupReader* _parent = nullptr;
```
Comment (DorianZheng) [inline be/src/formats/parquet/group_reader.h:114]: It's too much. Could you pass what is real needed.

File: be/src/formats/parquet/group_reader.cpp [hunk 4]
```diff
@@ -493,50 +433,125 @@ Status GroupReader::_lazy_skip_rows(const std::vector<int>& read_columns, const
...
+void GroupReader::DictFilterContext::use_as_dict_filter_column(int col_idx, SlotId slot_id,
+                                                               const std::vector<ExprContext*>& conjunct_ctxs) {
+    _is_dict_filter_column[col_idx] = true;
+    _dict_column_indices.emplace_back(col_idx);
+    _conjunct_ctxs_by_slot[slot_id] = conjunct_ctxs;
+}
 
-    auto iter = _dict_filter_preds.begin();
+Status GroupReader::DictFilterContext::rewrite_conjunct_ctxs_to_predicates() {
+    const GroupReaderParam& _param = _parent->_param;
```
Comment (DorianZheng) [inline be/src/formats/parquet/group_reader.cpp:451]: pass param as a const reference

File: be/src/formats/parquet/group_reader.cpp [hunk 5]
```diff
@@ -493,50 +433,125 @@ Status GroupReader::_lazy_skip_rows(const std::vector<int>& read_columns, const
...
+    for (int col_idx : _dict_column_indices) {
+        const auto& column = _param.read_cols[col_idx];
+        SlotId slot_id = column.slot_id;
+        vectorized::ChunkPtr dict_value_chunk = std::make_shared<vectorized::Chunk>();
+        std::shared_ptr<vectorized::BinaryColumn> dict_value_column = vectorized::BinaryColumn::create();
+        dict_value_chunk->append_column(dict_value_column, slot_id);
+
+        RETURN_IF_ERROR(_parent->_column_readers[slot_id]->get_dict_values(dict_value_column.get()));
```
Comment (DorianZheng) [inline be/src/formats/parquet/group_reader.cpp:459]: pass column readers as a const reference

File: be/src/formats/parquet/group_reader.cpp [hunk 6]
```diff
@@ -493,50 +433,125 @@ Status GroupReader::_lazy_skip_rows(const std::vector<int>& read_columns, const
...
+        }
+        _parent->_obj_pool.add(_predicates[slot_id]);
+    }
+
+    return Status::OK();
+}
+
+void GroupReader::DictFilterContext::init_chunk(vectorized::ChunkPtr* chunk) {
+    // replace dict filter column
+    size_t chunk_size = _parent->_param.chunk_size;
```
Comment (DorianZheng) [inline be/src/formats/parquet/group_reader.cpp:495]: pass chunk size as a argument

File: be/src/formats/parquet/group_reader.cpp [hunk 7]
```diff
@@ -493,50 +433,125 @@ Status GroupReader::_lazy_skip_rows(const std::vector<int>& read_columns, const
...
+    for (int col_idx : _dict_column_indices) {
+        const auto& column = _parent->_param.read_cols[col_idx];
```
Comment (DorianZheng) [inline be/src/formats/parquet/group_reader.cpp:497]: pass read_cols as a argument

File: be/src/formats/parquet/group_reader.cpp [hunk 8]
```diff
@@ -493,50 +433,125 @@ Status GroupReader::_lazy_skip_rows(const std::vector<int>& read_columns, const
...
         slot_id = iter->first;
         pred = iter->second;
         pred->evaluate_and((*chunk)->get_column_by_slot_id(slot_id).get(), filter->data());
     }
 }
 
-Status GroupReader::_dict_decode(vectorized::ChunkPtr* chunk) {
+Status GroupReader::DictFilterContext::decode_chunk(const vectorized::ChunkPtr& read_chunk,
+                                                    vectorized::ChunkPtr* chunk) {
+    const GroupReaderParam& _param = _parent->_param;
```
Comment (DorianZheng) [inline be/src/formats/parquet/group_reader.cpp:522]: ditto

File: be/src/formats/parquet/group_reader.cpp [hunk 9]
```diff
@@ -493,50 +433,125 @@ Status GroupReader::_lazy_skip_rows(const std::vector<int>& read_columns, const
...
+            vectorized::ColumnPtr& dict_codes = read_chunk->get_column_by_slot_id(slot_id);
+            vectorized::ColumnPtr& dict_values = (*chunk)->get_column_by_slot_id(slot_id);
+            dict_values->resize(0);
+
+            auto* codes_nullable_column =
+                    vectorized::ColumnHelper::as_raw_column<vectorized::NullableColumn>(dict_codes);
+            auto* codes_column = vectorized::ColumnHelper::as_raw_column<vectorized::FixedLengthColumn<int32_t>>(
+                    codes_nullable_column->data_column());
+            RETURN_IF_ERROR(
+                    _parent->_column_readers[slot_id]->get_dict_values(codes_column->get_data(), dict_values.get()));
```
Comment (DorianZheng) [inline be/src/formats/parquet/group_reader.cpp:541]: ditto

**Review by DorianZheng — CHANGES_REQUESTED** [review]

**Review by DorianZheng — APPROVED** [review]

**Review by stephen-shelby — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Review by Youngwb — APPROVED** [review]

**Pull request merged**